mbuild/lattice.py

```diff
@@ -0,0 +1,401 @@
+            if y is None:
+                y = 1
+            if z is None:
+                z = 1
+            if x < 1 or y < 1 or z < 1:
+                raise ValueError('Incorrect populate value: X, Y, or Z is < 1.'
+                                 ' Cannot replicate unit cell less than 1')
+        elif self.dimension == 2:
+            a = self.lattice_spacings[0]
+            b = self.lattice_spacings[1]
+            if x is None:
+                x = 1
+            if y is None:
+                y = 1
+            if z is None:
+                pass
+            else:
+                raise ValueError('Z is defined although dimension is 2D')
+            if x < 1 or y < 1:
+                raise ValueError('Incorrect populate value: X or Y is < 1. '
+                                 ' Cannot replicate unit cell less than 1')
+        elif self.dimension == 1:
+            a = self.lattice_spacings[0]
+            if x is None:
+                x = 1
+            if y is None:
+                pass
+            else:
+                raise ValueError('Y is defined although dimension is 1D')
+            if z is None:
+                pass
+            if z is not None:
+                raise ValueError('Z is defined although dimension is 2D')
+            if x < 1:
+                raise ValueError('Incorrect populate value: X < 1. '
+                                 ' Cannot replicate unit cell less than 1')
+        else:
+            raise ValueError('Dimension not defined.')
+
+        cell = defaultdict(list)
+        for key, val in self.basis_vectors.items():
+            for val_item in range(len(val)):
+                if self.dimension == 3:
+                    for i in range(x):
+                        for j in range(y):
+                            for k in range(z):
+                                tmpx = (val[val_item][0] + i) * a
+                                tmpy = (val[val_item][1] + j) * b
+                                tmpz = (val[val_item][2] + k) * c
+                                tmp_tuple = tuple((tmpx, tmpy, tmpz))
+                                cell[key].append(((tmp_tuple)))
+                elif self.dimension == 2:
+                    for i in range(x):
+                        for j in range(y):
+                            tmpx = (val[val_item][0] + i) * a
+                            tmpy = (val[val_item][1] + j) * b
+                            tmp_tuple = tuple((tmpx, tmpy))
+                            cell[key].append(((tmp_tuple)))
+                else:
+                    for i in range(x):
+                        tmpx = (val[val_item][0] + i) * a
+                        tmp_tuple = tuple((tmpx))
+                        cell[key].append(((tmp_tuple)))
+
+        ret_lattice = mb.Compound()
+        if compound_dict is None:
+            for key, val in cell.items():
+                tmp_part = mb.Particle(name=key, pos=[0, 0, 0])
+                for i in range(len(val)):
```

Comment:
There is almost always a nicer way to express this in Python.

E.g.

```python
for i, value in enumerate(val):
```

but better yet, if you do not NEED the index, avoid using it to begin with:

```python
for item in val:
    ...
    mb.translate(particle_to_add, list(item))
```